README.md
# What?

PySight2MISP is a project that can be run to be used as glue between [iSight intel API](https://docs.fireeye.com/iSight/index.html) and [MISP API](https://github.com/CIRCL/PyMISP).

# Status

This script is not maintaned - if you want to use it, there might be a need to fix a lot of stuff before, but still happy to accept PR.

# Why

To get indicators, IOCs etc. from FireEye iSight into a MISP system.

# How

    see INSTALL.md

It is recommended to have a "Publisher" account on the MISP system. A "sync user" account is not sufficient because it 
doesn't have enough rights (e.g. for tagging events).

You do not need an iSight account to test the script, there is test data to test with, e.g., a MISP VM.

# Roadmap

## Short term

* Fix [issues](issues).
* Add written report details per event.

## Long term

* Migrate stuff to a MISP import module.

CHANGELOG.md
PySight 1.3 (2020-05-08)
===========================
- Improvements in Threading.
- Don't log Pymisp output in debug mode to avoid disclosure of authorization keys.
- Remove simplejson from the requirements as it isn't used. Instead, list datetime as a requirement.

PySight 1.2 (2020-03-27)
===========================
- Add a comment in `README.md` regarding required access rights in MISP.
- Add a configuration option for the logging level and clean up logging.
- Bug fix to set the date when iSight doesn't provide a date.
- FireEye iSight reports can contain multiple ThreatScapes. This is taken into account now.
- For the malware family in iSight reports, use the "antivirus detection" category in MISP instead of 
"payload installation".
- When iSight provides an IP address in combination with a port, we always assume that it's a destination IP address.

PySight 1.1.2 (2020-02-24)
===========================
- Tagging of attributes didn't work yet. This bug has been fixed.
- Updated the example config file.
- Tried to make test_pysight.py work again, but didn't succeed.

PySight 1.1.1 (2020-02-14)
===========================
- Don't create filename attributes when the file's name is UNKNOWN (case insensitive).
- Improvements to comments of attributes and objects.
- Don't create ip address attributes when the IP address is provided in addition to a hostname.
- Remove the function to test connectivity to the FireEye iSight API.
- Create hostname|port attributes if both values are provided by FireEye iSight.
- Fix a mistake in adding comments to MISP objects.
- Remove the option to sleep in non-threaded processing.

PySight 1.1 (2020-02-11)
===========================
- Add configuration option for number of threads.
- Intercept response code 204 from FireEye iSight API.
- Differentiate whether a proxy shall be used for the FireEye iSight API, for MISP, or for both.
- Proper handling of publishDate data in pySightReport.py.
- Replace PyMISP with ExpandedPyMISP.
- Expand mapping of iSight fields to MISP objects, attributes and tags.

PySight 1.0.10 (2019-08-07)
===========================
- remove proxymanager

PySight 1.0.10 (2017-09-02)
===========================
- added Tag for C2 Attribute based

PySight 1.0.9 (2017-04-03)
===========================
- fixed date issue with reports (strings instead of float values in seconds)

PySight 1.0.8 (2017-01-25)
===========================
- started with attribute level tagging
- made it python3 ready
- adjusted the requirements
- added *.log to gitignore

PySight 1.0.7 (2016-11-25)
===========================
- intro of requirements.txt
- added some stuff to example config
- refactored a lot of variables
- removed a lot of issues
- added threading option and sleep time to config
- added C2 as bool in PySightReport
- added C2 check in File to add tags in the future as well
- added C2 example in test_data
- added other example file from https://docs.fireeye.com/iSight/index.html#/report_download
- removed threat actor other

PySight 1.0.6 (2016-11-04)
===========================
- initial version to be used with prod MISP
- fixed time / date issue

PySight 0.0.6 (2016-10-31)
===========================
- introducing test cases
- code cleaning
- new signatures for some methods

PySight 0.0.5 (2016-09-29)
===========================
- disable log messages from the Requests library
- remove file size temporary to reduce noise in an event

PySight 0.0.4 (2016-09-27)
===========================
- https://git.gcert.basf.net/fireeye/PySight/issues/3
- https://git.gcert.basf.net/fireeye/PySight/issues/5

PySight 0.0.3 (2016-09-26)
===========================
- modified pySightReport
- error handling
- better correlation
- improved parsing a lot
- pushed first sets to prod MISP
- logging to file
- saving of each indicator set to a file

PySight 0.0.2 (2016-09-21)
===========================
- added pySightReport.py
- added Changelog
- API with iSight is working

PySight 0.0.1 (2016-09-20)
===========================
init


test_pysight.py
import unittest

import PySight_settings
from PySight import misp_process_isight_indicators, data_search_report, test_isight_connection, misp_delete_events, \
    process_isight_indicator, get_misp_instance, isight_search_indicators


class OnlineTestCases(unittest.TestCase):
    def test_is_isight_working(self):
        """Is test api there aka connection test"""
        self.assertTrue(test_isight_connection())


    def test_get_isight_reports_last(self):
        import time
        since = int(time.time()) - 720 * 60 * 60
        print (since)
        return_value = isight_search_indicators(base_url=PySight_settings.isight_url,
                                                public_key=PySight_settings.isight_pub_key,
                                                private_key=PySight_settings.isight_priv_key, hours=720)
        import json
        print(json.dumps(return_value, sort_keys=True, indent=4))
        self.assertIsNotNone(return_value)


    def test_get_24hours_isight_report_and_push_misp(self):
        import time
        since = int(time.time()) - 720 * 60 * 60
        print(since)
        return_value = isight_search_indicators(base_url=PySight_settings.isight_url,
                                                public_key=PySight_settings.isight_pub_key,
                                                private_key=PySight_settings.isight_priv_key, hours=720)
        import json
        print(json.dumps(return_value, sort_keys=True, indent=4))

        misp_process_isight_indicators(return_value)

        self.assertIsNotNone(return_value)



    def test_is_misp_working(self):
        localmisp = get_misp_instance()
        self.assertIsNotNone(localmisp)

    def test_get_specific_report2(self):
        localmisp = get_misp_instance()
        event = localmisp.get_event('2355')
        for temp in event['Event']['Attribute']:
            if temp['id'] == '34539':
                attribute = temp
                break
        import json
        print(json.dumps(event, sort_keys=True, indent=4))
        print (event['Event']['Attribute'])
        localmisp.add_tag(attribute, 'PAP:GREEN', attribute=True)
        localmisp.add_tag(event, 'PAP:GREEN', attribute=False)


    def test_get_specific_report(self):
        localmisp = get_misp_instance()
        result = data_search_report(url=PySight_settings.isight_url,private_key=PySight_settings.isight_priv_key,public_key=PySight_settings.isight_pub_key,a_reportid="16-00017916")
        misp_process_isight_indicators(result)


        #result = data_search_indicators_last24_h(PySight_settings.isight_url, PySight_settings.isight_pub_key,
        #                                         PySight_settings.isight_priv_key)


class OfflineTestCases(unittest.TestCase):
    def test_parse_example_indicator(self):
        import json
        json_data = json.load(open("test_data/example_indicator.json"))
        PySight_settings.logger.debug(json.dumps(json_data['message'], sort_keys=True, indent=4))
        #for i in json_data['message']:
        returnvalue = process_isight_indicator(json_data['message'])
        #logger.debug("hols %s",json.dumps(i, sort_keys=True, indent=4))

        #returnvalue = pySightReport(json_data['message'])
        #logger.debug("Return valuer: %s",returnvalue)

        #self.assertIsNot(returnvalue,None,"Message received and not none")
        self.assertIsNot(returnvalue,False,"Message is false, something went wrong")

    def test_parse_example_report(self):
        import json
        json_data = json.load(open("test_data/example_report.json"))
        PySight_settings.logger.debug(json.dumps(json_data['message'], sort_keys=True, indent=4))
        # for i in json_data['message']:
        returnvalue = process_isight_indicator(json_data['message'])
        # logger.debug("hols %s",json.dumps(i, sort_keys=True, indent=4))

        # returnvalue = pySightReport(json_data['message'])
        # logger.debug("Return valuer: %s",returnvalue)

        # self.assertIsNot(returnvalue,None,"Message received and not none")
        self.assertIsNot(returnvalue, False, "Message is false, something went wrong")

    def test_parse_example_report_from_struart(self):
        import json
        json_data = json.load(open("test_data/real/16-00014704.json"))
        PySight_settings.logger.debug(json.dumps(json_data['report'], sort_keys=True, indent=4))
        # for i in json_data['message']:
        returnvalue = process_isight_indicator(json_data['report'])
        # logger.debug("hols %s",json.dumps(i, sort_keys=True, indent=4))

        # returnvalue = pySightReport(json_data['message'])
        # logger.debug("Return valuer: %s",returnvalue)

        # self.assertIsNot(returnvalue,None,"Message received and not none")
        self.assertIsNot(returnvalue, False, "Message is false, something went wrong")


    def test_parse_example_indicator_c2(self):
        import json
        json_data = json.load(open("test_data/example_indicator_c2.json"))
        #logger.debug(json.dumps(json_data['message'], sort_keys=True, indent=4))

        # for i in json_data['message']:
        returnvalue = process_isight_indicator(json_data['message'])
        # logger.debug("hols %s",json.dumps(i, sort_keys=True, indent=4))

        # returnvalue = pySightReport(json_data['message'])
        # logger.debug("Return valuer: %s",returnvalue)

        # self.assertIsNot(returnvalue,None,"Message received and not none")
        self.assertIsNot(returnvalue, False, "Message is false, something went wrong")

    # Not a good test case
    #def test_find_previous_events(self):
    #    self.assertIsNot(check_misp_all_result("16-00011458"),False)



    # def test_is_error_catched_with_wrong_credentials(self):
    #    print("asd")

    # def test_data_text_search_filter(self):
    #    self.assertIsNot(data_text_search_filter(isight_url, isight_pub_key, isight_priv_key), False)

    def test_delete_misp_events(self):
        localmisp = get_misp_instance()
        self.assertFalse(misp_delete_events(1519,1519,localmisp))
        self.assertTrue(misp_delete_events(1519,1519,localmisp))


    #def test_search_ip(self):
    #    data_basic_search_ip(isight_url, isight_pub_key, isight_priv_key, '8.8.8.8')
    #    self.assertTrue(data_basic_search_ip(isight_url, isight_pub_key, isight_priv_key,'8.8.8.8'))


if __name__ == '__main__':
    unittest.main()

model/network.py
"""
"network": [
                    {
                        "domain": "in-travelusa.com",
                        "actorId": "73c877bb-54fb-4f0a-90f8-c250bc303bcc",
                        "protocol": "HTTP",
                        "actor": "FIN7",
                        "networkType": "C&C",
                        "identifier": "Attacker"
                    },
"""

class iSightNetwork:
    def __init__(self,a_network_json):
        self.domain = None
        self.actorId = None
        self.protocol = None
        self.actor = None
        self.networkType = None
        self.identifier = None
        self._parse_json(a_network_json)

    def _parse_json(self, a_network_json):
        if not a_network_json:
            raise ValueError('No Json given')

        if 'domain' in a_network_json:
            self.domain = str(a_network_json['domain'])

        if 'actorId' in a_network_json:
            self.actorId = str(a_network_json['actorId'])

        if 'protocol' in a_network_json:
            self.protocol = str(a_network_json['protocol'])

        if 'actor' in a_network_json:
            self.actor = str(a_network_json['actor'])

        if 'identifier' in a_network_json:
            self.identifier = str(a_network_json['identifier'])

        if 'networkType' in a_network_json:
            self.networkType = str(a_network_json['networkType'])


        # Print out the Json given to the method
        # logger.debug(json.dumps(p_alert_json, sort_keys=False, indent=4, separators=(',', ': ')))

PySight_settings.py
import configparser
import logging


# Initialize the config parser
config = configparser.RawConfigParser()

# Read the config file and set config values
config.read('config.cfg')

LOG_LEVEL = config.get('general', 'log_level')
use_threading = config.getboolean('general', 'use_threading')
# If threading is requested, set the number of threads according to the configuration.
# Otherwise only use one thread.
if use_threading:
    number_threads = config.getint('general', 'number_threads')
else:
    number_threads = 1

isight_url = config.get('isight', 'isight_url')
isight_priv_key = config.get('isight', 'isight_priv_key')
isight_pub_key = config.get('isight', 'isight_pub_key')
isight_last_hours = config.getint('isight', 'last_hours')

misp_url = config.get('MISP', 'misp_url')
misp_key = config.get('MISP', 'misp_key')
misp_verifycert = config.getboolean('MISP', 'misp_verifycert')


USE_ISIGHT_PROXY = config.getboolean('proxy', 'use_isight_proxy')
USE_MISP_PROXY = config.getboolean('proxy', 'use_misp_proxy')
if USE_ISIGHT_PROXY or USE_MISP_PROXY:
    PROXY_HOST = config.get('proxy', 'host')
    PROXY_PORT = config.get('proxy', 'port')
    PROXY_PROTOCOL = config.get('proxy', 'protocol')
    proxy = config.get('proxy', 'full')
    proxy_address = proxy

debug_mode = False

# Create a logger.
logger = logging.getLogger(__name__)
# Set the loglevel, also for imported modules.
if LOG_LEVEL.upper() == 'DEBUG':
    logger.setLevel(logging.DEBUG)
    # Pymisp DEBUG logging would include the authorization key which we want to avoid
    logging.getLogger('pymisp').setLevel(logging.INFO)
    logging.getLogger('urllib3').setLevel(logging.DEBUG)
    debug_mode = True
elif LOG_LEVEL.upper() == 'INFO':
    logger.setLevel(logging.INFO)
    logging.getLogger('pymisp').setLevel(logging.INFO)
    logging.getLogger('urllib3').setLevel(logging.INFO)
elif LOG_LEVEL.upper() == 'WARNING':
    logger.setLevel(logging.WARNING)
    logging.getLogger('pymisp').setLevel(logging.WARNING)
    logging.getLogger('urllib3').setLevel(logging.WARNING)
elif LOG_LEVEL.upper() == 'ERROR':
    logger.setLevel(logging.ERROR)
    logging.getLogger('pymisp').setLevel(logging.ERROR)
    logging.getLogger('urllib3').setLevel(logging.ERROR)
elif LOG_LEVEL.upper() == 'CRITICAL':
    logger.setLevel(logging.CRITICAL)
    logging.getLogger('pymisp').setLevel(logging.CRITICAL)
    logging.getLogger('urllib3').setLevel(logging.CRITICAL)
else:
    print('Invalid logging level "%s". Using default level WARNING.' % LOG_LEVEL)
    logger.setLevel(logging.WARNING)
    logging.getLogger('pymisp').setLevel(logging.WARNING)
    logging.getLogger('urllib3').setLevel(logging.WARNING)

# Create a file handler and log there, too, in addition to the console.
log_file = logging.FileHandler('output.log')
formatter = logging.Formatter('%(asctime)s - %(name)s - %(levelname)s - %(message)s')
log_file.setFormatter(formatter)
logger.addHandler(log_file)


model/pySightReport.py
#!/usr/bin/env python

# pySightReport - Python class for parsing iSightReport json alerts
#
# Alexander Jaeger (deralexxx)
#
# The MIT License (MIT) see https://github.com/deralexxx/FireMISP/blob/master/LICENSE
#
# For documentation of iSight indicator fields, see
# https://docs.fireeye.com/iSight/index.html#/field_definitions and
# https://docs.fireeye.com/iSight/index.html#/indicators
#

from datetime import datetime
import logging
# Regular expressions are not used yet.
# Potentially they might be required to split registry fields into hive, key and value.
#import re
import time

#reload(sys)
#sys.setdefaultencoding('utf-8')

# Initialize the logger
logging.basicConfig(level=logging.DEBUG)
logger = logging.getLogger(__name__)
formatter = logging.Formatter('%(asctime)s - %(name)s - %(levelname)s - %(message)s')


class pySightReport (object):
    def __init__(self, a_alert_json):
        """
        :param a_alert_json:
        :type a_alert_json:
        :rtype: object
        """

        # Initialize all potential fields of the FireEye iSight report
        self.alert = a_alert_json

        # General information
        self.reportId = None
        self.title = None  # Cutwail Botnet Distributes Recruitment Mass Mailings
        self.publishDate = None  # 1469544180
        self.ThreatScape = None  # Cyber Espionage
        self.riskRating = None  # High
        self.audience = None  # Operational
        self.intelligenceType = None  # threat / malware / vulnerability / overview
        self.reportLink = None  # https:#api.isightpartners.com/report/16-00011458
        self.webLink = None  # https:#mysight.isightpartners.com/report/full/16-00011458

        # Email-related indicators
        self.emailIdentifier = None  # Attacker
        self.senderAddress = None  # lissddzz@gmail.com
        self.senderName = None  # lissddzz
        self.sourceDomain = None  # samyongonc.com
        self.sourceIP = None  # 184.105.137.110
        self.subject = None  # Administrator Manager position
        self.recipient = None  # yeh@cwb.gov.tw
        self.emailLanguage = None  # English

        # File-related indicators
        self.fileName = None  # TW2BBFF500.doc
        self.fileSize = None  # 14860
        self.fuzzyHash = None  # 384:AmHWrWG6qqrx7F7ByIvjgS+S0SBS1n9dwnHJWNy/4yAOksmSfsF17BtX7K4:qrWG6qqV7F7ByIvjgS+S0SBSd9dwnHJw
        self.fileIdentifier = None  # Related
        self.md5 = None  # d27eb3f18ba7f3ae6fa793630882652f
        self.sha1 = None  # 4559ba637772b681dee07127c7c17c776455138e
        self.sha256 = None  # e9c60a120db8a4366734dcecbc15ddd4510ef7929cc7a5d21529180494a35cdc
        self.description = None  # Keylogger
        self.fileType = None  # Rich Text Format data, version 1, ANSI
        self.packer = None  # Armadillo v1.xx - v2.xx
        self.registry = None  # HKEY_LOCAL_MACHINE\SOFTWARE\CBSTEST
        self.registryHive = None  # HKEY_LOCAL_MACHINE
        self.registryKey = None  # Software\Microsoft\Windows\CurrentVersion\RunOnce
        self.registryValue = None  # SilentApp
        self.fileCompilationDateTime = None  # 1371573858
        self.filePath = None  # /tmp/adversary/test

        # Network-related indicators
        self.userAgent = None  # Mozilla
        self.asn = None  # 26272
        self.cidr = None  # 1.179.132.0/24
        self.domain = None  # webmonder.gicp.net
        self.domainTimeOfLookup = None  # 1371573858
        self.networkIdentifier = None  # Attacker
        self.ip = None  # 112.121.182.148
        self.port = None  # 80
        self.protocol = None  # TCP
        self.registrantEmail = None  # sammyguy@gmail.com
        self.registrantName = None  # Vanella Salvatore
        self.url = None  # http://www.google.com
        self.networkType = None  # C&C

        # Context
        self.malwareFamily = None  # Dyre
        self.malwareFamilyId = None  # 42bceb96-13c5-4b3f-a435-92ad9b17db27
        self.actor = None  # actor-oldms
        self.actorId = None  # c9fc4d46-516b-4fdb-8272-2798d2cdf0bd
        self.observationTime = None  # 1469544180

        # After initialization, parse the report and assign all available values
        self._parse_json(a_alert_json)


    def _parse_json(self, p_alert_json):
        """
        :param p_alert_json:
        :type p_alert_json:
        """
        if not p_alert_json:
            raise ValueError('No Json given')

        if "reportId" in p_alert_json and p_alert_json['reportId'] is not None:
            self.reportId = str(p_alert_json['reportId'])
        if 'title' in p_alert_json and p_alert_json['title'] is not None:
            self.title = str(p_alert_json['title'])
        if 'publishDate' in p_alert_json and p_alert_json['publishDate'] is not None:
            if isinstance(p_alert_json['publishDate'],int):
                self.publishDate = p_alert_json['publishDate']
            # If publishDate is not in epoch format, i.e. not an integer, we suppose it to be human readable,
            # e.g. "October 11, 2016 07:20:00 AM", and convert it to integer epoch format
            else:
                logger.debug('Converting date %s to an epoch format', p_alert_json['publishDate'])
                date_format = '%B %d, %Y %H:%M:%S %p'
                datetime_object = datetime.strptime(p_alert_json['publishDate'], date_format)
                timestamp = time.mktime(datetime_object.timetuple())
                self.publishDate = int(timestamp)
        if 'ThreatScape' in p_alert_json and p_alert_json['ThreatScape'] is not None:
            self.ThreatScape = str(p_alert_json['ThreatScape'])
        if 'audience' in p_alert_json and p_alert_json['audience'] is not None:
            self.audience = str(p_alert_json['audience'])
        if 'intelligenceType' in p_alert_json and p_alert_json['intelligenceType'] is not None:
            self.intelligenceType = str(p_alert_json['intelligenceType'])
        if 'reportLink' in p_alert_json and p_alert_json['reportLink'] is not None:
            self.reportLink = str(p_alert_json['reportLink'])
        if 'webLink' in p_alert_json and p_alert_json['webLink'] is not None:
            self.webLink = str(p_alert_json['webLink'])
            # Split it to get the ma_id of the "alert-url"
            # ("https://fireeye.foo.bar/event_stream/events_for_bot?ma_id=12345678")
            #self.alert_ma_id = (self.alert_url.split("="))[1]
        if 'emailIdentifier' in p_alert_json and p_alert_json['emailIdentifier'] is not None:
            self.emailIdentifier = str(p_alert_json['emailIdentifier'])
        if 'senderAddress' in p_alert_json and p_alert_json['senderAddress'] is not None:
            self.senderAddress = str(p_alert_json['senderAddress'])
        if 'senderName' in p_alert_json and p_alert_json['senderName'] is not None:
            self.senderName = str(p_alert_json['senderName'])
        if 'sourceDomain' in p_alert_json and p_alert_json['sourceDomain'] is not None:
            self.sourceDomain = str(p_alert_json['sourceDomain'])
        if 'sourceIP' in p_alert_json and p_alert_json['sourceIP'] is not None:
            self.sourceIP = str(p_alert_json['sourceIP'])
        if 'subject' in p_alert_json and p_alert_json['subject'] is not None:
            self.subject = str(p_alert_json['subject'])
        if 'recipient' in p_alert_json and p_alert_json['recipient'] is not None:
            self.recipient = str(p_alert_json['recipient'])
        if 'emailLanguage' in p_alert_json and p_alert_json['emailLanguage'] is not None:
            self.emailLanguage = str(p_alert_json['emailLanguage'])
        if 'fileName' in p_alert_json and p_alert_json['fileName'] is not None:
            self.fileName = str(p_alert_json['fileName'])
        if 'fileSize' in p_alert_json and p_alert_json['fileSize'] is not None:
            self.fileSize = str(p_alert_json['fileSize'])
        if 'fuzzyHash' in p_alert_json and p_alert_json['fuzzyHash'] is not None:
            self.fuzzyHash = str(p_alert_json['fuzzyHash'])
        if 'fileIdentifier' in p_alert_json and p_alert_json['fileIdentifier'] is not None:
            self.fileIdentifier = str(p_alert_json['fileIdentifier'])
        if 'md5' in p_alert_json and p_alert_json['md5'] is not None:
            self.md5 = str(p_alert_json['md5'])
        if 'sha1' in p_alert_json and p_alert_json['sha1'] is not None:
            self.sha1 = str(p_alert_json['sha1'])
        if 'sha256' in p_alert_json and p_alert_json['sha256'] is not None:
            self.sha256 = str(p_alert_json['sha256'])
        if 'description' in p_alert_json and p_alert_json['description'] is not None:
            self.description = str(p_alert_json['description'])
        if 'fileType' in p_alert_json and p_alert_json['fileType'] is not None:
            self.fileType = str(p_alert_json['fileType'])
        if 'packer' in p_alert_json and p_alert_json['packer'] is not None:
            self.packer = str(p_alert_json['packer'])
        # TODO: Ideally, the registry field would be separated into hive, key and value.
        if 'registry' in p_alert_json and p_alert_json['registry'] is not None:
            self.registry = str(p_alert_json['registry'])
        if 'registryHive' in p_alert_json and p_alert_json['registryHive'] is not None:
            self.registryHive = str(p_alert_json['registryHive'])
        if 'registryKey' in p_alert_json and p_alert_json['registryKey'] is not None:
            self.registryKey = str(p_alert_json['registryKey'])
        if 'registryValue' in p_alert_json and p_alert_json['registryValue'] is not None:
            self.registryValue = str(p_alert_json['registryValue'])
        if 'fileCompilationDateTime' in p_alert_json and p_alert_json['fileCompilationDateTime'] is not None:
            if isinstance(p_alert_json['fileCompilationDateTime'], int):
                self.fileCompilationDateTime = p_alert_json['fileCompilationDateTime']
            # If fileCompilationDateTime is not in epoch format, i.e. not an integer, we suppose it to be human
            # readable, e.g. "October 11, 2016 07:20:00 AM", and convert it to integer epoch format.
            else:
                logger.debug('Converting timestamp %s to epoch fomat', p_alert_json['fileCompilationDateTime'])
                date_format = '%B %d, %Y %H:%M:%S %p'
                datetime_object = datetime.strptime(p_alert_json['fileCompilationDateTime'], date_format)
                timestamp = time.mktime(datetime_object.timetuple())
                self.fileCompilationDateTime = int(timestamp)
        if 'filePath' in p_alert_json and p_alert_json['filePath'] is not None:
            self.filePath = str(p_alert_json['filePath'])
        if 'userAgent' in p_alert_json and p_alert_json['userAgent'] is not None:
            self.userAgent = str(p_alert_json['userAgent'])
        if 'asn' in p_alert_json and p_alert_json['asn'] is not None:
            self.asn = str(p_alert_json['asn'])
        if 'cidr' in p_alert_json and p_alert_json['cidr'] is not None:
            self.cidr = str(p_alert_json['cidr'])
        if 'domain' in p_alert_json and p_alert_json['domain'] is not None:
            self.domain = str(p_alert_json['domain'])
        if 'domainTimeOfLookup' in p_alert_json and p_alert_json['domainTimeOfLookup'] is not None:
            if isinstance(p_alert_json['domainTimeOfLookup'], int):
                self.domainTimeOfLookup = p_alert_json['domainTimeOfLookup']
            # If domainTimeOfLookup is not in epoch format, i.e. not an integer, we suppose it to be human readable,
            # e.g. "October 11, 2016 07:20:00 AM", and convert it to integer epoch format
            else:
                logger.debug('Converting timestamp %s to epoch format', p_alert_json['domainTimeOfLookup'])
                date_format = '%B %d, %Y %H:%M:%S %p'
                datetime_object = datetime.strptime(p_alert_json['domainTimeOfLookup'], date_format)
                timestamp = time.mktime(datetime_object.timetuple())
                self.domainTimeOfLookup = int(timestamp)
        if 'networkIdentifier' in p_alert_json and p_alert_json['networkIdentifier'] is not None:
            self.networkIdentifier = str(p_alert_json['networkIdentifier'])
        if 'ip' in p_alert_json and p_alert_json['ip'] is not None:
            self.ip = str(p_alert_json['ip'])
        if 'port' in p_alert_json and p_alert_json['port'] is not None:
            self.port = str(p_alert_json['port'])
        if 'protocol' in p_alert_json and p_alert_json['protocol'] is not None:
            self.protocol = str(p_alert_json['protocol'])
        if 'registrantEmail' in p_alert_json and p_alert_json['registrantEmail'] is not None:
            self.registrantEmail = str(p_alert_json['registrantEmail'])
        if 'registrantName' in p_alert_json and p_alert_json['registrantName'] is not None:
            self.registrantName = str(p_alert_json['registrantName'])
        if 'url' in p_alert_json and p_alert_json['url'] is not None:
            self.url = str(p_alert_json['url'])
        if 'networkType' in p_alert_json and p_alert_json['networkType'] is not None:
            self.networkType = str(p_alert_json['networkType'])
        if 'malwareFamily' in p_alert_json and p_alert_json['malwareFamily'] is not None:
            self.malwareFamily = str(p_alert_json['malwareFamily'])
        if 'malwareFamilyId' in p_alert_json and p_alert_json['malwareFamilyId'] is not None:
            self.malwareFamilyId = str(p_alert_json['malwareFamilyId'])
        if 'actor' in p_alert_json and p_alert_json['actor'] is not None:
            self.actor = str(p_alert_json['actor'])
        if 'actorId' in p_alert_json and p_alert_json['actorId'] is not None:
            self.actorId = str(p_alert_json['actorId'])
        if 'observationTime' in p_alert_json and p_alert_json['observationTime'] is not None:
            if isinstance(p_alert_json['observationTime'], int):
                self.observationTime = p_alert_json['observationTime']
            # If observationTime is not in epoch format, i.e. not an integer, we suppose it to be human readable,
            # e.g. "October 11, 2016 07:20:00 AM", and convert it to integer epoch format
            else:
                logger.debug('Converting timestamp %s to epoch format', p_alert_json['observationTime'])
                date_format = '%B %d, %Y %H:%M:%S %p'
                datetime_object = datetime.strptime(p_alert_json['observationTime'], date_format)
                timestamp = time.mktime(datetime_object.timetuple())
                self.observationTime = int(timestamp)
        if 'riskRating' in p_alert_json and p_alert_json['riskRating'] is not None:
            self.riskRating = str(p_alert_json['riskRating'])

        logger.debug('Finished parsing %s', p_alert_json)
